### PR TITLE
feat: soft deletion model and reindexing algolia job memory improvements

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
## Description

Description of changes made

## Ticket Link

Link to the associated ticket
[Link title](https://openedx.atlassian.net/browse/ENT-XXXX)

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

**Review by johnnagro — APPROVED** [review]

File: enterprise_catalog/apps/api/tasks.py [hunk 1]
```diff
@@ -486,112 +552,100 @@ def index_content_keys_in_algolia(content_keys, algolia_client):  # pylint: disa
...
+            ).all().iterator()
+
+        catalog_mapping_values_length = sum((len(catalogs) for catalogs in catalog_uuids_by_key.values()))
+        customer_mapping_values_length = sum((len(customers) for customers in customer_uuids_by_key.values()))
+        queries_mapping_values_length = sum((len(queries) for queries in catalog_queries_by_key.values()))
+        logger.info(
+            f'[ENTERPRISE_CATALOG_ALGOLIA_REINDEX] content mappings values lengths: catalog_uuids_by_key- '
+            f'{catalog_mapping_values_length}, customer_uuids_by_key- {customer_mapping_values_length}, '
+            f'catalog_queries_by_key- {queries_mapping_values_length}'
+        )
```
Comment (iloveagent57) [inline enterprise_catalog/apps/api/tasks.py]: Hmm, I'm actually not sure this is useful if the batch size is 1.  We'd just end up with a couple thousand INFO lines saying how many catalogs/customers/queries a single piece of content is associated with.

File: enterprise_catalog/apps/api/tasks.py [hunk 2]
```diff
@@ -486,112 +552,100 @@ def index_content_keys_in_algolia(content_keys, algolia_client):  # pylint: disa
...
+            Q(content_metadata__content_key__in=content_keys_batch)
+            | Q(content_metadata__parent_content_key__in=content_keys_batch)
+            | Q(
+                content_metadata__associated_content_metadata__content_key__in=content_keys_batch,
+                content_metadata__content_type=PROGRAM
+            )
         )
+        all_memberships = ContentMetadataToQueries.objects.select_related(
+            'catalog_query', 'content_metadata'
+        ).filter(query).all().iterator()
```
Comment (iloveagent57) [inline enterprise_catalog/apps/api/tasks.py]: `all()` doesn't make sense here.

**Review by iloveagent57 — COMMENTED** [review]
Maybe a more useful thing to log is the maximum number of uuids we stored in the mappings across all iterations of the main loop.

File: enterprise_catalog/apps/api/tasks.py (hunk 1 above)
Comment (alex-sheehan-edx) [reply enterprise_catalog/apps/api/tasks.py]: fair enough, I was just erring on the side of more information. Happy to remove

**Review by iloveagent57 — APPROVED** [review]

**Pull request merged**